index.js

```diff
@@ -129,47 +75,129 @@ app.use(function(req, res, next) {
 
-	const translators = JSON.parse(req.body.translators);
-	const firstChunkOnly =
+	res.translators = JSON.parse(req.body.translators);
+
+	if (fromCache) {
+		return getFile(`${res.uuid}_${res.translators[0]}`)
+			.then(data => res.json(data))
+			.catch(err => console.log(err));
+	}
+	
+	res.firstChunkOnly =
 		!req.query.hasOwnProperty('firstChunkOnly') || !!req.query.firstChunkOnly; // default is firstChunkOnly=true
 
-	CAPI.get(uuid)
+	CAPI.get(res.uuid)
 		.then(async data => {
 			const text = data.bodyXML;
-			const title = maybeAppendDot(data.title);
-			let standfirst = '';
+			const title = Utils.maybeAppendDot(data.title);
+			res.standfirst = '';
 			if (data.standfirst) {
-				standfirst = maybeAppendDot(data.standfirst);
+				res.standfirst = Utils.maybeAppendDot(data.standfirst);
 			} // adding a closing . improves the translation
-			const combinedText = title + '\n\n' + standfirst + '\n\n' + text;
+			res.combinedText = title + '\n\n' + res.standfirst + '\n\n' + text;
+			res.pubDate = data.publishedDate;
 
-			generateTranslations(
-				translators,
-				combinedText,
-				lang,
-				firstChunkOnly,
-				standfirst
-			).then(translations => {
-				translations.article = uuid;
-				res.json(translations);
-			});
+			if(checkCache) {
+				const promises = [];
+				for(let i = 0; i < res.translators.length; ++i) {
+					const check = CACHE.checkAndGet(`${res.uuid}_${res.translators[i]}`, res.lang.toLowerCase(), res.pubDate);
+					promises.push(check);
+				}
+				
+				return Promise.all(promises)
+					.then(data => {
+						if(data.length !== 1 && data.every( item => item === data[0] )) {
+							return next();
+						}
+
+						const names = ['original'].concat(res.translators);
+						const newTranslations = [];
+						const cachedTranslations = [];
+
+						let formattedResult = {
+							article: res.uuid,
+							texts: {
+								'original': extract(res.combinedText)
+							},
+							translatorNames: names,
+							audioUrls: {},
+							audioButtonText: {}
+						}
+
+						for(let i = 0; i < data.length; ++i) {
+							if(!data[i]) {
+								newTranslations.push(res.translators[i]);
+							} else {
+								cachedTranslations.push(res.translators[i]);
+								formattedResult.texts[res.translators[i]] = JSON.parse(data[i])[res.lang.toLowerCase()];
+							}
+						}
+
+						formattedResult = Utils.formatOutput(formattedResult, !!res.standfirst, true, cachedTranslations.concat('original'));
+						formattedResult = Audio.get(formattedResult, res.lang, cachedTranslations.concat('original'));
+
+						if(newTranslations.length > 0) {
+							if(cachedTranslations.length > 0) {
+								res.translators = newTranslations;
+								res.formattedResult = formattedResult;
+							}
+							return next();
+						} else {
+							return res.json(formattedResult);
+						}
+					})
+					.catch(err => console.log('CHECK cache error', err));
+					
+			} else {
+				return next();
+			}
```

Comment:
Is the index file getting a bit bloated? Maybe we could break it down into smaller files.

Reply:
Yes, I took big chunks out already. It could do with further refactor but at this point, I'm keen to get something out. What is everybody thinking? /cc @Sqyphen @railsagainstignorance